Objective-C/CBLProperties.mm

```diff
@@ -146,101 +144,346 @@ - (BOOL) containsObjectForKey: (NSString*)key {
 
 
 - (nullable id) objectForKeyedSubscript: (NSString*)key {
-    if (_properties)
-        return _properties[key];
-    else
-        return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+    id obj = _properties ? _properties[key] : nil;
+    if (!obj) {
+        auto value = [self fleeceValueForKey: key];
+        if (value == nullptr)
+            return nil;
+        obj = [self fleeceValueToObject: value  forKey: key asJson: NO];
+        [self cacheValue: obj forKey: key changed: NO];
+    }
+    return obj;
 }
 
 
 - (void) setObject: (nullable id)value forKeyedSubscript: (NSString*)key {
-    [self mutateProperties];
-    
-    // NSDate:
-    if ([value isKindOfClass: NSDate.class])
-        value = [CBLJSON JSONObjectWithDate: value];
-    
-    id oldValue = _properties[key];
-    
+    id oldValue = self[key];
     if (![value isEqual: oldValue] && value != oldValue) {
-        [_properties setValue: value forKey: key];
+        value = [self convertValue: value oldValue: oldValue forKey: key];
+        [self cacheValue: value forKey: key changed: YES];
         [self markChanges];
     }
 }
 
 
-#pragma mark - INTERNAL:
+- (void) revert {
+    for (NSString* key in _changedNames) {
+        id value = _properties[key];
+        if ([value isKindOfClass: [CBLSubdocument class]]) {
+            CBLSubdocument* subdoc = $cast(CBLSubdocument, value);
```

Comment:
`$cast` will do another isKindOf: test, which is redundant; just do a C cast instead. (Or do a `$castIf` and assign it to subdoc, then test if subdoc is nil; that's the pattern I usually use.)